File: 116_populatingNextRightPointersInEachNode.cpp
// Time Complexity : O(N)
// Space Complexity :O(N)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

class Solution {
public:
    Node* connect(Node* root) {
        if(root == NULL)
            return root;
        queue<Node*> q;
        q.push(root);
        while(!q.empty()){
            int size = q.size();
            Node* prev = q.front();
            q.pop();
            if(prev->left != NULL){
                q.push(prev->left);
                q.push(prev->right);
            }
            for(int i=1; i<size; i++){
                Node* curr = q.front();
                q.pop();
                if(curr->left != NULL){
                    q.push(curr->left);
                    q.push(curr->right);
                }
                prev->next = curr;
                prev = curr;
            }
        }
        return root;
    }
};

// Time Complexity : O(N)
// Space Complexity :O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no
class Solution {
public:
    Node* connect(Node* root) {
        if(root == NULL)
            return root;
        Node* level = root;
        while(level->left != NULL){
            Node* curr = level;
            while(curr != NULL){
                curr->left->next = curr->right;
                if(curr->next != NULL)
                    curr->right->next = curr->next->left;
                curr = curr->next;
            }
            level = level->left;
        }
        return root;
    }
};


// Time Complexity : O(N)
// Space Complexity :O(H)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no
class Solution {
public:
    Node* connect(Node* root) {
        if(root == NULL)
            return root;
        dfs(root->left, root->right);
        return root;
    }
    
    void dfs(Node* left, Node* right){
        //base
        if(left == NULL)
            return;
        //logic
        left->next = right;
        dfs(left->left, left->right);
        dfs(left->right, right->left);
        dfs(right->left, right->right);
    }
};

File: _116_PopulatingNextRightPointersinEachNode.java
// Time Complexity : o(n) where n is number of nodes in tree
// Space Complexity : o(h) whre h is the height of tree
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
// pass parent along with each node, use parent to find the next node of each child
class Solution {
    public Node connect(Node root) {
        preorder(root, null);
        return root;
    }

    public void preorder(Node root, Node parent){

        if(root == null) return;

        Node temp = parent;
        while(temp != null){

            if(temp.left != parent.left && temp.left != null){
                root.next = temp.left;
                break;
            }
            else if(root != temp.right && temp.right != null){
                root.next = temp.right;
                break;
            }
            temp = temp.next;
        }

        preorder(root.left, root);
        preorder(root.right, root);
    }

    /******** sort of same as above, instead of passing parent to the given node, pass its next node and attach the next to the current node
       calculate the next node for left and rght chidren and pass it in recursion *****/
       public Node connect(Node root) {
        preorder(root, null);
        return root;
    }

    public void preorder(Node root, Node next){
        if(root == null) return;
        root.next = next;
        preorder(root.left, root.right);
        preorder(root.right, root.next != null ? root.next.left : null);
    }

    /*** one more logic of maintaining two levele pointers and keep attaching next of current level using prev level pointer***/
    // being used in LC - 117, curr moves 1 time and prev moves 1 time as curr level 2 the number of nodes as prev level.  
}
